### ISSUE TITLE
Replication and Nodes in OpenStorage

### ISSUE THREAD

**Issue by WolfyLPDC** [libopenstorage/openstorage#689]
**Is this a BUG REPORT or FEATURE REQUEST?**:
Just a question. I know it is not the place but I sent emails to the info email and there is no response.

**What happens*:
What does replication mean in OpenStorage? As far as I understand, OS is just a layer that allows to bind a remote volume to a local filesystem though the OSD. However, in the API, at the response of the call _/v1/volumes/{volume_id}_, I can see the variable _replica_set_, which is, or I understand it is, a list of nodes where the volume is repeated. What does this mean? How can I set or enable it?

Also, when you run a new Open Storage Daemon, in the config file you can set the variables _clusterid_ and _nodeid_, so I get that there can be more than one cluster with more than one node each one, but I am trying to do tests with multiple OSDs and multiple configurations but _/v1/nodes_ always show only one node. I think that, perhaps, both questions are related.

Also, I would like to understand the general meaning. Why set replication and multiple nodes in a binding layer? Should not in be the job of the NFS Server, for example?

**Comment by lpabon** [comment]
Hi, I will try to answer the questions below:

>What does replication mean in OpenStorage? As far as I understand, OS is just a layer that allows to bind a remote volume to a local filesystem though the OSD. However, in the API, at the response of the call /v1/volumes/{volume_id}, I can see the variable replica_set, which is, or I understand it is, a list of nodes where the volume is repeated. What does this mean? How can I set or enable it?

Normally replicas is the number of copies for a given object/volume. This is handled by `VolumeSpec.ha_level`. `VolumeSpec.Replica_sets`, on the other hand, are used to determine the locations of each of the copies throughout the cluster. Note, that this is dependent on the driver implementation behind OpenStorage.

>Also, when you run a new Open Storage Daemon, in the config file you can set the variables clusterid and nodeid, so I get that there can be more than one cluster with more than one node each one, but I am trying to do tests with multiple OSDs and multiple configurations but /v1/nodes always show only one node. I think that, perhaps, both questions are related.

Are you pointing the osd program to the same KVDB(etcd or consul) database? Also, are you starting each osd program with its own node id but the same cluster id?
FYI, I am working on a CSI based NFS driver which will provide non-shared and shared volumes as well as raw block volumes. This will use a container running osd and setup to run as a daemonset in Kubernetes. Therefore, I will need to setup osd to talk to the same Kvdb, use the same cluster id, and take the node id from Kubernetes. More information can be found at issue #670 

>Also, I would like to understand the general meaning. Why set replication and multiple nodes in a binding layer? Should not in be the job of the NFS Server, for example?

The OpenStorage API is a general API, but it is up to the driver implementation to use the values in their own way. So, yes, it is the job of the NFS server to figure out what to do with replication. As far as I know, the NFS driver does nothing with ha_lavel or replica_sets since it does not support IO going to multiple servers. Instead, I may be wrong, but only the Portworx driver supports that.

Hope this answers your questions.

**Comment by WolfyLPDC** [comment]
Hi @lpabon 

First of all, thank you very much for the answer. I'm documenting some tools and I was very blocked at OpenStorage because of this. 

> Note, that this is dependent on the driver implementation behind OpenStorage.

When I was asking about replication in OpenStorage, my question was related about what handles replication, if OpenStorage or the storage system. The answer is pretty clear here.

> Are you pointing the osd program to the same KVDB(etcd or consul) database? Also, are you starting each osd program with its own node id but the same cluster id?

About this point... I was not using the same KVDB because when I did the tests for this I was not using KVDB at all. I though it was only useful to persist the volumes. Now, it sounds pretty obvious. 

> and take the node id from Kubernetes

Is there a way to set the node id without modifying the config file? There should be some kind of "--node-id" option in the OSD CLI. Also, what has to do this "clusterization"? For the NFS Server, at least, there is no difference between to independent clients and one cluster. The files are synced in the same way. Is it for the Portworx storage?

I'm not talking about your Kubernetes work because I do not know Kubernetes at all. I'm using Docker because it is what I am used to.

The last question is pretty well answered too.

Thank you so much.

**Comment by lpabon** [comment]
>  There should be some kind of "--node-id" option in the OSD CLI
Yes, I have a change on this because I am working on a CSI OpenStorage driver based on NFS.  See #691 

> Also, what has to do this "clusterization"? For the NFS Server, at least, there is no difference between to independent clients and one cluster. The files are synced in the same way. Is it for the Portworx storage?
Correct and any other driver implementation that is not stored in this repo.

> I'm not talking about your Kubernetes work because I do not know Kubernetes at all. I'm using Docker because it is what I am used to.

Well, Kubernetes is really awesome, you should try it some time. Try out [minikube](https://kubernetes.io/docs/setup/minikube/) just to get things started.

At some point I will have my CSI (which stands for Container Storage Interface) driver ready for NFS (not for production use :) ) and you can try it out in Kubernetes.

**Comment by WolfyLPDC** [comment]
Thank you very much for the info.

I'm currently looking for (and documenting) some useful storage-related tools to use together with Docker, so there is no way I can use Kubernetes. I would like to try/learn it but currently it is not my job. However, I appreciate the recommendation.

About the _--node-id_ param, i have seen that the pull request have been accepted. Good job.
